### DISCUSSION TITLE
Incorrect typing for Scope and Message in ASGIApp

### DISCUSSION THREAD

**Discussion by ods** [encode/starlette#2040] [Potential Issue]
According to [ASGI spec](https://asgi.readthedocs.io/en/latest/specs/main.html#applications) both scope and event message must be dictionaries, not an arbitrary mutable mapping:

> * `scope`: The connection scope information, a **dictionary** that contains at least a `type` key specifying the protocol that is incoming
> * `receive`: an awaitable callable that will yield a new event **dictionary** when one is available
> * `send`: an awaitable callable taking a single event **dictionary** as a positional argument that will return once the send has been completed or the connection has been closed

and in other place:

> `scope` must be a `dict`.

But in starlette [they are declared](https://github.com/encode/starlette/blob/master/starlette/types.py#L3-L4) as `MutableMapping`:

```python
Scope = typing.MutableMapping[str, typing.Any]
Message = typing.MutableMapping[str, typing.Any]

Receive = typing.Callable[[], typing.Awaitable[Message]]
Send = typing.Callable[[Message], typing.Awaitable[None]]

ASGIApp = typing.Callable[[Scope, Receive, Send], typing.Awaitable[None]]
```

This actually breaks interoperability with other libraries that follows the spec. For example, the following code

```python
from starlette.applications import Starlette
from httpx import ASGITransport

ASGITransport(Starlette())
```

gives the following error when type-checked with mypy:

```
error: Argument 1 to "ASGITransport" has incompatible type "Starlette"; expected "Callable[[Dict[str, Any], Callable[[], Awaitable[Dict[str, Any]]], Callable[[Dict[str, Any]], Coroutine[None, None, None]]], Coroutine[None, None, None]]"  [arg-type]
note: "Starlette.__call__" has type "Callable[[Arg(MutableMapping[str, Any], 'scope'), Arg(Callable[[], Awaitable[MutableMapping[str, Any]]], 'receive'), Arg(Callable[[MutableMapping[str, Any]], Awaitable[None]], 'send')], Coroutine[Any, Any, None]]"
```

The [typing of ASGI app in httpx](https://github.com/encode/httpx/blob/master/httpx/_transports/asgi.py#L17-L24) follows the spec:

```python
_Message = typing.Dict[str, typing.Any]
_Receive = typing.Callable[[], typing.Awaitable[_Message]]
_Send = typing.Callable[
    [typing.Dict[str, typing.Any]], typing.Coroutine[None, None, None]
]
_ASGIApp = typing.Callable[
    [typing.Dict[str, typing.Any], _Receive, _Send], typing.Coroutine[None, None, None]
]
```

**Comment by sk-** [comment]
Note that these broader types also generated issues with the latest version of [opentelemetry-python-contrib](https://github.com/open-telemetry/opentelemetry-python-contrib/issues/2565).

It would be great if the types could be fixed, or if the spec could be amended.

Maybe @Kludex could have some insights as he has driven some typing efforts, or @tomchristie who added said types.

**Comment by adriangb** [comment]
I don't think the issue is as simple as what you're proposing:
https://pyright-play.net/?strict=true&code=GYJw9gtgBALgngBwJYDsDmUkQWEMoCCKcANIQO4CGSMlARgDYCmZAwpQw-c2QLICutRk16UEydAChpAEybBYISigDOOPAAoxCAFxR2nbkwDaxmUgDGMYypggyROAF0nDqjSPGAcmBRMXAJRQALQAfFA%2BfjqSULFQAHSJ0pKUKnAoFlByChBwAPraGioWYAhMegJCzKLiqGg2dg7ETkFhEb7lMXGJ8dJ2ymq4MBq5BeIBkkA
https://mypy-play.net/?mypy=latest&python=3.12&gist=6e14a107050185893787deca37974b90

I see that you're using a [version of MyPy](https://github.com/open-telemetry/opentelemetry-python-contrib/blob/3511ed180d568fa180420b9fc4968cd8c70a94c7/dev-requirements.txt#L6) that is [3 years old](https://github.com/python/mypy/releases/tag/v0.930). Maybe try updating mypy or using pyright?

**Reply by ods** [reply]
Here is the results for the sample above, with `mypy 1.10.0`:
```
sample.py:4: error: Argument 1 to "ASGITransport" has incompatible type "Starlette"; expected "Callable[[dict[str, Any], Callable[[], Awaitable[dict[str, Any]]], Callable[[dict[str, Any]], Coroutine[None, None, None]]], Coroutine[None, None, None]]"  [arg-type]
sample.py:4: note: "Starlette.__call__" has type "Callable[[Arg(MutableMapping[str, Any], 'scope'), Arg(Callable[[], Awaitable[MutableMapping[str, Any]]], 'receive'), Arg(Callable[[MutableMapping[str, Any]], Awaitable[None]], 'send')], Coroutine[Any, Any, None]]"
Found 1 error in 1 file (checked 1 source file)
```
and `pyright 1.1.366`:
```
sample.py:4:15 - error: Argument of type "Starlette" cannot be assigned to parameter "app" of type "_ASGIApp" in function "__init__"
    Type "Starlette" is incompatible with type "_ASGIApp"
      Parameter 3: type "_Send" is incompatible with type "Send"
        Type "_Send" is incompatible with type "Send"
          Parameter 1: type "Message" is incompatible with type "Dict[str, Any]"
            "MutableMapping[str, Any]" is incompatible with "Dict[str, Any]" (reportArgumentType)
1 error, 0 warnings, 0 informations 
```
The error message from `pyright` suggests, that the problem is with `Send`, not `Scope`. Actually changing the signature of `Starlette.__call__` to accept `send: typing.Callable[[dict[str, typing.Any]], typing.Awaitable[None]]` seems to fix typing errors with both `mypy` and `pyright`.

**Comment by Kludex** [comment]
What is the proposed solution here?

**Reply by ods** [reply]
This depends on our goal. If we want to be fully ASGI compliant, then we have to switch to `dict` in type declarations as suggested in [pull request](https://github.com/encode/starlette/pull/2058/files). But… having httpx [doesn't follow the spec](https://github.com/encode/httpx/pull/3109) anymore too, just to be compatible with FastAPI… I already don't believe it has a chance to be fixed. Note, that the types are _not more permissive_ here and in httpx. They're just different from ASGI, so both starlette/FastAPI and httpx may have problems with ASGI-compliant code.
